packages/netstring/writer.js

```diff
@@ -1,51 +1,54 @@
 // @ts-check
 /// <reference types="ses"/>
 
-const COLON = ':'.charCodeAt(0);
-const COMMA = ','.charCodeAt(0);
+const COMMA_BUFFER = new Uint8Array([','.charCodeAt(0)]);
 
-// The initial buffer length should be small enough to not be an imposition in
-// the common case but large enough to avoid reallocation in the common case.
-// It also must be long enough to fit an ASCII decimal representation of the
-// length of the longest possible message.
-// This is a guess:
-const initialBufferLength = 128;
-
-const encoder = new TextEncoder();
+/** @param {number} length */
+const getLengthPrefixCharCodes = length =>
+  // eslint-disable-next-line no-bitwise
+  [...`${length | 0}:`].map(char => char.charCodeAt(0));
 
 /**
  * @param {import('@endo/stream').Writer<Uint8Array, undefined>} output
- * @returns {import('@endo/stream').Writer<Uint8Array, undefined>}
+ * @param {object} [opts]
+ * @param {boolean} [opts.chunked]
+ * @returns {import('@endo/stream').Writer<Uint8Array | Uint8Array[], undefined>}
  */
-export const makeNetstringWriter = output => {
+export const makeNetstringWriter = (output, { chunked = false } = {}) => {
   return harden({
-    async next(message) {
-      // Must allocate to support concurrent writes.
-      let buffer = new Uint8Array(initialBufferLength);
-      const { written: colonAt = 0 } = encoder.encodeInto(
-        `${message.byteLength}`,
-        buffer,
+    async next(messageChunks) {
+      if (!Array.isArray(messageChunks)) {
+        messageChunks = [messageChunks];
+      }
+
+      const messageLength = messageChunks.reduce(
+        (acc, { length }) => acc + length,
+        0,
       );
-      const messageAt = colonAt + 1;
-      const commaAt = messageAt + message.byteLength;
-      const messageLength = commaAt + 1;
 
-      // Grow buffer if necessary.
-      if (messageLength > buffer.byteLength) {
-        let newCapacity = buffer.byteLength;
-        while (newCapacity < messageLength) {
-          newCapacity *= 2;
-        }
-        const newBuffer = new Uint8Array(newCapacity);
-        newBuffer.set(buffer, 0);
-        buffer = newBuffer;
-      }
+      const prefix = getLengthPrefixCharCodes(messageLength);
 
-      buffer[colonAt] = COLON;
-      buffer.set(message, messageAt);
-      buffer[commaAt] = COMMA;
+      if (chunked) {
+        return Promise.all([
+          output.next(new Uint8Array(prefix)),
+          ...messageChunks.map(async chunk => output.next(chunk)),
+          output.next(COMMA_BUFFER),
```

Comment:
I’m not sure whether calling `next` after a final iteration has well-defined behavior. I don’t think this is wrong, but it might be good to document somewhere that this writer might write after the target closes.

Reply:
Yeah it's definitely an aggressive writer in chunk writing mode. Good thing it's not the default, and the node-stream writer can take it. I can add a note.

Reply:
Note added in the description of the maker.